File: week-4/subsets/subsets.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var subsets = function(nums) {
    const res = [];
    recurse(nums, 0, [], res);
    return res;
};

const recurse = (nums, index, cur_array, res) => {
    res.push(cur_array.slice());
    
    for (let i=index; i<nums.length; i++) {
        cur_array.push(nums[i]);
        recurse(nums, i+1, cur_array, res);
        cur_array.pop();
    }
};

File: week-4/subsets/subsets.py
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        res = []
        self.recurse(nums, 0, [], res)
        return res
    
    def recurse(self, nums, index, cur_array, res):
        res.append(cur_array[:])
        
        for i in range(index, len(nums)):
            cur_array.append(nums[i])
            self.recurse(nums, i+1, cur_array, res)
            cur_array.pop()
